pandas/_libs/parsers.pyx

```diff
@@ -2084,27 +2068,34 @@ cdef inline int _try_bool_flex_nogil(parser_t *parser, int64_t col,
     return 0
 
 
-cdef kh_str_t* kset_from_list(list values) except NULL:
+cdef kh_str_starts_t* kset_from_list(list values) except NULL:
     # caller takes responsibility for freeing the hash table
     cdef:
         Py_ssize_t i
         khiter_t k
-        kh_str_t *table
+        kh_str_starts_t *table
         int ret = 0
 
         object val
 
-    table = kh_init_str()
+    table = kh_init_str_starts()
 
     for i in range(len(values)):
         val = values[i]
 
         # None creeps in sometimes, which isn't possible here
         if not isinstance(val, bytes):
-            kh_destroy_str(table)
+            kh_destroy_str_starts(table)
             raise ValueError('Must be all encoded bytes')
 
-        k = kh_put_str(table, PyBytes_AsString(val), &ret)
+        kh_put_str_starts_item(table, PyBytes_AsString(val), &ret)
+
+    if table.table.n_buckets <= 128:
+        # Resize the hash table to make it almost empty, this
+        # reduces amount of hash collisions on lookup thus
+        # "key not in table" case is faster.
+        # Note that this trades table memory footprint for lookup speed.
```

Comment:
Is there a ballpark estimate for how much more memory this might be using? I feel if anything we get more issues for read_csv in regards to high memory usage than slow parsing so just want to be careful of tradeoffs that we make, especially since we don't have any memory benchmarks set up

Reply:
Worst case would be when `bucket_size` was increased from 128 to 1024.
Memory for 128-sized table (on x86_64) ~3.6 KB, for 1024-sized table (on x86_64) ~21.04 KB.

For default case bucket size is 32, so memory is increased from 1.67 KB to 6.04 KB.

I think it's acceptable to trade about 15 more KB (as there are typically three such maps - NA values, true-ish values, false-ish values) to gain up to 15% of parsing speed.

P.S. Memory size formula where `B` is `bucket_size`:
```c++
sizeof(khint32_t) * 4 + sizeof(void*) * 3 + B * (sizeof(khint32_t) + sizeof(char*) + sizeof(size_t)) + 256 * sizeof(int)
```
which gives `1064 + B * 20` bytes  (again on x86_64).

Reply:
My math might be fuzzy but assuming someone is parsing say 10M records and this increases the size of each record by 15kb, would that require roughly 150MB more of memory?

Reply:
It is 15 KB _per parser_, not per record :)
So in a typical case (when one parser exists at a time) it's only 15 KB more.

Reply:
Ha thanks for clarifying. I suppose that's quite the difference